File: ios/App/App/FolioleCompanionSyncPackEnvelopeValidator.swift
import CryptoKit
import Foundation
import ZIPFoundation

struct FoliolePreparedSyncPack {
    let databaseBytes: Data
    let manifest: [String: Any]
    let rowCounts: [String: Int]
}

enum FolioleCompanionSyncPackEnvelopeValidator {
    private static let sqliteHeader = Data("SQLite format 3\0".utf8)

    static func validate(
        archiveURL: URL,
        contract: FolioleCompanionSyncPackContract,
        expectedPeerId: String,
        expectedSourcePeerId: String
    ) throws -> FoliolePreparedSyncPack {
        do {
            let entries = try readEntries(archiveURL, contract: contract)
            let manifest = try parseManifest(entries["manifest.json"])
            let rowCounts = try validateManifest(
                manifest,
                contract: contract,
                expectedPeerId: expectedPeerId,
                expectedSourcePeerId: expectedSourcePeerId
            )
            guard let compressed = entries[contract.databaseEntry] else { throw invalid("missing_sync_pack_entry") }
            try verifySha256(compressed, expected: try string(manifest, "database_compressed_sha256"), layer: "compressed")
            let database = try FolioleCompanionZlib.inflate(compressed)
            try verifySha256(database, expected: try string(manifest, "database_uncompressed_sha256"), layer: "uncompressed")
            guard database.starts(with: sqliteHeader) else { throw invalid("invalid_sync_pack_sqlite_header") }
            return FoliolePreparedSyncPack(databaseBytes: database, manifest: manifest, rowCounts: rowCounts)
        } catch let error as NSError where error.domain == "FolioleSyncPack" {
            throw error
        } catch {
            throw invalid("invalid_sync_pack_container", cause: error)
        }
    }

    private static func readEntries(
        _ url: URL,
        contract: FolioleCompanionSyncPackContract
    ) throws -> [String: Data] {
        let archive = try Archive(url: url, accessMode: .read)
        let allowed = Set(["manifest.json", contract.databaseEntry])
        var result: [String: Data] = [:]
        for entry in archive {
            guard entry.type == .file, allowed.contains(entry.path) else { throw invalid("invalid_sync_pack_entry") }
            guard result[entry.path] == nil else { throw invalid("duplicate_sync_pack_entry") }
            var data = Data()
            _ = try archive.extract(entry) { data.append($0) }
            result[entry.path] = data
        }
        guard Set(result.keys) == allowed else { throw invalid("missing_sync_pack_entry") }
        return result
    }

    private static func parseManifest(_ data: Data?) throws -> [String: Any] {
        guard let data else { throw invalid("missing_sync_pack_entry") }
        let value = try JSONSerialization.jsonObject(with: data)
        guard let manifest = value as? [String: Any] else { throw invalid("invalid_sync_pack_manifest") }
        return manifest
    }

    private static func validateManifest(
        _ manifest: [String: Any],
        contract: FolioleCompanionSyncPackContract,
        expectedPeerId: String,
        expectedSourcePeerId: String
    ) throws -> [String: Int] {
        guard try string(manifest, "format") == contract.format else { throw invalid("unsupported_sync_pack_format") }
        guard try integer(manifest, "format_version") == contract.formatVersion else {
            throw invalid("unsupported_sync_pack_format_version")
        }
        guard try string(manifest, "compression") == contract.compression else {
            throw invalid("unsupported_sync_pack_compression")
        }
        guard try string(manifest, "database_file") == contract.databaseEntry else {
            throw invalid("invalid_sync_pack_database_entry")
        }
        let schema = try integer(manifest, "schema_version")
        guard (contract.minimumSchemaVersion...contract.maximumSchemaVersion).contains(schema) else {
            throw invalid("unsupported_sync_pack_schema_version")
        }
        guard try string(manifest, "to_peer_id") == expectedPeerId else { throw invalid("sync_pack_target_mismatch") }
        _ = try string(manifest, "pack_id")
        guard try string(manifest, "from_device_id") == expectedSourcePeerId else {
            throw invalid("sync_pack_source_mismatch")
        }
        _ = try string(manifest, "created_at")
        let from = try integer(manifest, "from_state_seq")
        let to = try integer(manifest, "to_state_seq")
        guard from >= 0, to >= from else { throw invalid("invalid_sync_pack_state_range") }
        return try tableCounts(manifest, required: contract.manifestTableNames)
    }

    static func tableCounts(_ manifest: [String: Any], required: Set<String>? = nil) throws -> [String: Int] {
        guard let tables = manifest["tables"] as? [[String: Any]] else {
            throw invalid("invalid_sync_pack_table_manifest")
        }
        var result: [String: Int] = [:]
        for table in tables {
            let name = try string(table, "name")
            let count = try integer(table, "row_count")
            guard count >= 0, result[name] == nil, required?.contains(name) != false else {
                throw invalid("invalid_sync_pack_table_manifest")
            }
            result[name] = count
        }
        if let required, Set(result.keys) != required { throw invalid("invalid_sync_pack_table_manifest") }
        return result
    }

    static func string(_ object: [String: Any], _ key: String) throws -> String {
        guard let value = object[key] as? String, !value.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            throw invalid("invalid_sync_pack_manifest_field")
        }
        return value.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    static func integer(_ object: [String: Any], _ key: String) throws -> Int {
        guard let value = object[key] as? NSNumber,
              CFGetTypeID(value) != CFBooleanGetTypeID(), value.doubleValue == Double(value.intValue) else {
            throw invalid("invalid_sync_pack_manifest_field")
        }
        return value.intValue
    }

    private static func verifySha256(_ data: Data, expected: String, layer: String) throws {
        let digest = SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
        guard "sha256:\(digest)" == expected else { throw invalid("invalid_sync_pack_\(layer)_checksum") }
    }

    static func invalid(_ code: String, cause: Error? = nil) -> NSError {
        var info: [String: Any] = [NSLocalizedDescriptionKey: code]
        if let cause { info[NSUnderlyingErrorKey] = cause }
        return NSError(domain: "FolioleSyncPack", code: 1, userInfo: info)
    }
}

File: android/app/src/main/java/com/foliole/android/FolioleCompanionSyncPackEnvelopeValidator.java
package com.foliole.android;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Set;
import java.util.zip.InflaterInputStream;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;

final class FolioleCompanionSyncPackEnvelopeValidator {
    private static final int COPY_BUFFER_BYTES = 256 * 1024;
    private static final byte[] SQLITE_HEADER = "SQLite format 3\0".getBytes(StandardCharsets.US_ASCII);

    private FolioleCompanionSyncPackEnvelopeValidator() {}

    static PreparedEnvelope validate(
        byte[] body,
        FolioleCompanionSyncPackContract contract,
        String expectedPeerId,
        String expectedSourcePeerId
    ) throws Exception {
        try {
            Map<String, byte[]> entries = readEntries(body, contract);
            JSONObject manifest = new JSONObject(new String(
                requireEntry(entries, "manifest.json"),
                StandardCharsets.UTF_8
            ));
            Map<String, Integer> rowCounts = validateManifest(
                manifest, contract, expectedPeerId, expectedSourcePeerId
            );
            byte[] compressed = requireEntry(entries, contract.databaseEntry());
            verifySha256(compressed, requireString(manifest, "database_compressed_sha256"), "compressed");
            byte[] database = inflate(compressed);
            verifySha256(database, requireString(manifest, "database_uncompressed_sha256"), "uncompressed");
            if (!hasSqliteHeader(database)) {
                throw invalid("invalid_sync_pack_sqlite_header");
            }
            return new PreparedEnvelope(contract, database, manifest, rowCounts);
        } catch (IllegalArgumentException exception) {
            throw exception;
        } catch (Exception exception) {
            throw new IllegalArgumentException("invalid_sync_pack_container", exception);
        }
    }

    private static Map<String, byte[]> readEntries(
        byte[] body,
        FolioleCompanionSyncPackContract contract
    ) throws Exception {
        Map<String, byte[]> entries = new LinkedHashMap<>();
        Set<String> allowed = new java.util.LinkedHashSet<>();
        allowed.add("manifest.json");
        allowed.add(contract.databaseEntry());
        try (ZipInputStream input = new ZipInputStream(new ByteArrayInputStream(body))) {
            ZipEntry entry;
            while ((entry = input.getNextEntry()) != null) {
                String name = entry.getName();
                if (entry.isDirectory() || !allowed.contains(name)) {
                    throw invalid("invalid_sync_pack_entry");
                }
                if (entries.put(name, readAll(input)) != null) {
                    throw invalid("duplicate_sync_pack_entry");
                }
                input.closeEntry();
            }
        }
        if (!entries.keySet().equals(allowed)) {
            throw invalid("missing_sync_pack_entry");
        }
        return entries;
    }

    private static Map<String, Integer> validateManifest(
        JSONObject manifest,
        FolioleCompanionSyncPackContract contract,
        String expectedPeerId,
        String expectedSourcePeerId
    ) throws Exception {
        if (!contract.format().equals(requireString(manifest, "format"))) {
            throw invalid("unsupported_sync_pack_format");
        }
        if (requireInt(manifest, "format_version") != contract.formatVersion()) {
            throw invalid("unsupported_sync_pack_format_version");
        }
        if (!contract.compression().equals(requireString(manifest, "compression"))) {
            throw invalid("unsupported_sync_pack_compression");
        }
        if (!contract.databaseEntry().equals(requireString(manifest, "database_file"))) {
            throw invalid("invalid_sync_pack_database_entry");
        }
        int schemaVersion = requireInt(manifest, "schema_version");
        if (schemaVersion < contract.minimumSchemaVersion() || schemaVersion > contract.maximumSchemaVersion()) {
            throw invalid("unsupported_sync_pack_schema_version");
        }
        if (!expectedPeerId.equals(requireString(manifest, "to_peer_id"))) {
            throw invalid("sync_pack_target_mismatch");
        }
        requireString(manifest, "pack_id");
        if (!expectedSourcePeerId.equals(requireString(manifest, "from_peer_id"))) {
            throw invalid("sync_pack_source_mismatch");
        }
        requireString(manifest, "created_at");
        int fromStateSeq = requireInt(manifest, "from_state_seq");
        int toStateSeq = requireInt(manifest, "to_state_seq");
        if (fromStateSeq < 0 || toStateSeq < fromStateSeq) {
            throw invalid("invalid_sync_pack_state_range");
        }
        return validateTableManifest(manifest.getJSONArray("tables"), contract.manifestTableNames());
    }

    private static Map<String, Integer> validateTableManifest(JSONArray tables, Set<String> required) throws Exception {
        Map<String, Integer> counts = new LinkedHashMap<>();
        for (int index = 0; index < tables.length(); index += 1) {
            Object item = tables.get(index);
            if (!(item instanceof JSONObject)) throw invalid("invalid_sync_pack_table_manifest");
            JSONObject table = (JSONObject) item;
            String name = requireString(table, "name");
            int count = requireInt(table, "row_count");
            if (count < 0 || !required.contains(name) || counts.put(name, count) != null) {
                throw invalid("invalid_sync_pack_table_manifest");
            }
        }
        if (!counts.keySet().equals(required)) throw invalid("invalid_sync_pack_table_manifest");
        return counts;
    }

    private static String requireString(JSONObject object, String key) throws Exception {
        Object value = object.get(key);
        if (!(value instanceof String) || ((String) value).trim().isEmpty()) {
            throw invalid("invalid_sync_pack_manifest_field");
        }
        return ((String) value).trim();
    }

    private static int requireInt(JSONObject object, String key) throws Exception {
        Object value = object.get(key);
        if (!(value instanceof Number)) throw invalid("invalid_sync_pack_manifest_field");
        double number = ((Number) value).doubleValue();
        int integer = ((Number) value).intValue();
        if (number != integer) throw invalid("invalid_sync_pack_manifest_field");
        return integer;
    }

    private static byte[] inflate(byte[] bytes) throws Exception {
        try (InflaterInputStream input = new InflaterInputStream(new ByteArrayInputStream(bytes))) {
            return readAll(input);
        } catch (Exception exception) {
            throw new IllegalArgumentException("invalid_sync_pack_compressed_database", exception);
        }
    }

    private static void verifySha256(byte[] bytes, String expected, String layer) throws Exception {
        if (!sha256Uri(bytes).equals(expected)) {
            throw invalid("invalid_sync_pack_" + layer + "_checksum");
        }
    }

    private static String sha256Uri(byte[] bytes) throws Exception {
        byte[] hash = MessageDigest.getInstance("SHA-256").digest(bytes);
        StringBuilder result = new StringBuilder("sha256:");
        for (byte value : hash) result.append(String.format("%02x", value));
        return result.toString();
    }

    private static byte[] requireEntry(Map<String, byte[]> entries, String name) {
        byte[] bytes = entries.get(name);
        if (bytes == null || bytes.length == 0) throw invalid("missing_sync_pack_entry");
        return bytes;
    }

    private static byte[] readAll(java.io.InputStream input) throws Exception {
        ByteArrayOutputStream output = new ByteArrayOutputStream();
        byte[] buffer = new byte[COPY_BUFFER_BYTES];
        int read;
        while ((read = input.read(buffer)) >= 0) output.write(buffer, 0, read);
        return output.toByteArray();
    }

    private static boolean hasSqliteHeader(byte[] bytes) {
        if (bytes.length < SQLITE_HEADER.length) return false;
        for (int index = 0; index < SQLITE_HEADER.length; index += 1) {
            if (bytes[index] != SQLITE_HEADER[index]) return false;
        }
        return true;
    }

    private static IllegalArgumentException invalid(String code) {
        return new IllegalArgumentException(code);
    }

    static final class PreparedEnvelope {
        final FolioleCompanionSyncPackContract contract;
        final byte[] databaseBytes;
        final JSONObject manifest;
        final Map<String, Integer> rowCounts;

        PreparedEnvelope(
            FolioleCompanionSyncPackContract contract,
            byte[] databaseBytes,
            JSONObject manifest,
            Map<String, Integer> rowCounts
        ) {
            this.contract = contract;
            this.databaseBytes = databaseBytes;
            this.manifest = manifest;
            this.rowCounts = rowCounts;
        }
    }
}
